torch_np/_ufuncs.py

```diff
@@ -146,6 +146,49 @@ def matmul(
     return result
 
 
+# ldexp casting is special : the dtype of the result == dtype of the 1st arg
+@normalizer
+def ldexp(
+    x1: Union[ArrayLike, Scalar],
+    x2: Union[ArrayLike, Scalar],
+    /,
+    out: Optional[OutArray] = None,
+    *,
+    where: NotImplementedType = True,
+    casting: Optional[CastingModes] = "same_kind",
+    order: NotImplementedType = "K",
+    dtype: Optional[DTypeLike] = None,
+    subok: NotImplementedType = False,
+    signature: NotImplementedType = None,
+    extobj: NotImplementedType = None,
+):
+
+    if dtype is not None:
+        if isinstance(x1, torch.Tensor):
+            x1 = _util.typecast_tensor(x1, dtype, casting)
+        else:
+            x1 = torch.as_tensor(x1, dtype=dtype)
+    else:
+        if not isinstance(x1, torch.Tensor):
+            x1 = torch.as_tensor(x1)
+        if _dtypes_impl._category(x1.dtype) < 2:
+            # cast integers to float64
+            x1 = x1.double()
```

Comment:
`.double()` is actually wrong because we allow for user control of  the default dtypes, so it needs to be `default_dtypes.float_dtype`. Ditto for unary ufuncs, and I've extracted the relevant three-liner into `_util` in the updated PR. Does that count as a reuse in your books? :-).